testing/fil_token_integration/actors/basic_nft_actor/src/lib.rs

```diff
@@ -55,6 +62,60 @@ fn invoke(params: u32) -> u32 {
             sdk::sself::set_root(&cid).unwrap();
             NO_DATA_BLOCK_ID
         }
+        "Approve" => {
+            let params = deserialize_params::<ApproveParams>(params);
+            handle.approve(&caller_address(), &params.operator, &params.token_ids).unwrap();
+            let cid = handle.flush().unwrap();
+            sdk::sself::set_root(&cid).unwrap();
+            NO_DATA_BLOCK_ID
+        }
+        "Revoke" => {
+            let params = deserialize_params::<RevokeParams>(params);
+            handle.revoke(&caller_address(), &params.operator, &params.token_ids).unwrap();
+            let cid = handle.flush().unwrap();
+            sdk::sself::set_root(&cid).unwrap();
+            NO_DATA_BLOCK_ID
+        }
+        "ApproveForAll" => {
+            let params = deserialize_params::<ApproveForAllParams>(params);
+            handle.approve_for_owner(&caller_address(), &params.operator).unwrap();
+            let cid = handle.flush().unwrap();
+            sdk::sself::set_root(&cid).unwrap();
+            NO_DATA_BLOCK_ID
+        }
+        "RevokeForAll" => {
+            let params = deserialize_params::<RevokeForAllParams>(params);
+            handle.revoke_for_all(&caller_address(), &params.operator).unwrap();
+            let cid = handle.flush().unwrap();
+            sdk::sself::set_root(&cid).unwrap();
+            NO_DATA_BLOCK_ID
+        }
+        "Transfer" => {
+            let params = deserialize_params::<TransferParams>(params);
+            handle.transfer(
+                &caller_address(),
+                &params.to,
+                &params.token_ids,
+                params.operator_data,
+                RawBytes::default()
+            ).unwrap();
+            let cid = handle.flush().unwrap();
+            sdk::sself::set_root(&cid).unwrap();
```

Comment:
If this calls a receiver hook and flushes state/sets root before that, it won't need to do so here. If it's not doing that, it needs to, otherwise the the receiver hook (which could make further calls to the token actor) will only have the previous state to work with, which will have no record of the transfer.

Looks like minting is doing the right thing already and transfers will follow?

Reply:
Yeh, I'm testing the right structure with minting first and will do Transfer/TransferFrom in a followup